src/GraphQL/Internal/Validation.hs

```diff
@@ -183,18 +205,42 @@ validateArguments args = mapErrors DuplicateArgument (makeMap [(name, value) | A
 -- imply that everything is valid.
 
 -- | A GraphQL selection.
-data Selection spread
-  = SelectionField (Field spread)
+data Selection' spread
+  = SelectionField (Field' spread)
   | SelectionFragmentSpread spread
   | SelectionInlineFragment (InlineFragment spread)
   deriving (Eq, Show)
 
+-- | Get all of the fields directly inside the given selection set.
+--
+-- TODO: This ignores fragments, whereas it should actually do something with
+-- them.
+--
+-- TODO: At this point, we ought to know that field names are unique. As such,
+-- we should return an ordered map of Name to Fields, rather than a bland
+-- list.
+getFields :: SelectionSet -> [Field]
+getFields ss = [field | SelectionField field <- ss]
+
 -- | A field in a selection set, which itself might have children which might
 -- have fragment spreads.
-data Field spread
-  = Field (Maybe Alias) Name Arguments Directives [Selection spread]
+data Field' spread
+  = Field' (Maybe Alias) Name Arguments Directives [Selection' spread]
   deriving (Eq, Show)
 
+instance HasName (Field' spread) where
+  getName (Field' _ name _ _ _) = name
+
+type Field = Field' FragmentSpread
+
+-- | Get the arguments of a field.
+getArguments :: Field -> Arguments
+getArguments (Field' _ _ args _ _) = args
```

Comment:
Essentially, yes. 

My thinking is that it seems a shame to go to all the work of constructively proven valid objects (and in many cases, the type signature does constitute a proof of validity), and then let just any bozo construct any random thing and pretend it's valid.

Secondarily, the data types are still in flux. A forthcoming PR will change the validation data types to do variable validation and selection set validation, and another will do variable substitution, fragment handling, and directive processing. Providing functions like these give a useful choke point for making those changes quickly without having to reason globally. "What does the resolver want? Oh, it only needs the arguments for this thing. I'll have it ask for those and I'll figure out how to provide them." 

View patterns actually might be a good fit for this module. I looked into them, for what we're doing at the moment, they are lots more typing.

FWIW, the most important data hiding functions are `getFields` and `getFieldSelectionSets`, which control _traversal_, which is qualitatively different from getting an attribute. (e.g. `getFields` has logic in it, and may well continue to do so).